### PR TITLE
PlayerAuthInput protocol updates

### PR DESCRIPTION
**Pull request by Prim69**

### CODE REVIEW

File: src/PlayerAuthInputPacket.php [hunk 1]
```diff
@@ -159,9 +232,28 @@ protected function encodePayload(PacketSerializer $out) : void{
...
+					$changedSlot->write($out);
+				}
+			}
+			$this->itemInteractionData->encode($out);
+		}
+		if($this->hasFlag(PlayerAuthInputFlags::PERFORM_ITEM_STACK_REQUEST)){
+			$this->itemStackRequest->write($out);
+		}
+		if($this->hasFlag(PlayerAuthInputFlags::PERFORM_BLOCK_ACTIONS)){
+			$out->putVarInt(count($this->blockActions));
```
Comment (dktapps) [inline src/PlayerAuthInputPacket.php:249]: are you sure? Usually lengths are unsigned.

File: src/types/PlayerBlockAction.php [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+	public const PREDICT_DESTROY = 26;
+	public const CONTINUE = 27;
+
+	public int $actionType;
+	public BlockPosition $blockPos;
+	public int $face;
+
+	public function read(PacketSerializer $in) : void {
+		$this->actionType = $in->getVarInt();
+		if(in_array($this->actionType, [PlayerBlockAction::ABORT_BREAK, PlayerBlockAction::START_BREAK, PlayerBlockAction::CRACK_BREAK, PlayerBlockAction::PREDICT_DESTROY, PlayerBlockAction::CONTINUE])){
```
Comment (dktapps) [inline src/types/PlayerBlockAction.php:43]: Please don't use in_array() ... use match() or if/else

File: src/types/PlayerBlockAction.php [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+			$this->blockPos = $in->getBlockPosition();
+			$this->face = $in->getVarInt();
+		}
+	}
+
+	public function write(PacketSerializer $out) : void {
+		$out->putVarInt($this->actionType);
+		if(in_array($this->actionType, [PlayerBlockAction::ABORT_BREAK, PlayerBlockAction::START_BREAK, PlayerBlockAction::CRACK_BREAK, PlayerBlockAction::PREDICT_DESTROY, PlayerBlockAction::CONTINUE])){
```
Comment (dktapps) [inline src/types/PlayerBlockAction.php:51]: Same again here.

**Comment by Prim69** [comment]
Fixed all that

File: src/types/PlayerBlockAction.php [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+	public const PREDICT_DESTROY = 26;
+	public const CONTINUE = 27;
+
+	public int $actionType;
+	public BlockPosition $blockPos;
+	public int $face;
+
+	public function read(PacketSerializer $in) : void{
+		$this->actionType = $in->getVarInt();
+		if(match ($this->actionType) {
```
Comment (alvin0319) [inline src/types/PlayerBlockAction.php:42]:
```suggestion
		if(match($this->actionType){
```

File: src/types/PlayerBlockAction.php [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+			default => false
+		}){
+			$this->blockPos = $in->getBlockPosition();
+			$this->face = $in->getVarInt();
+		}
+	}
+
+	public function write(PacketSerializer $out) : void{
+		$out->putVarInt($this->actionType);
+		if(match ($this->actionType) {
```
Comment (alvin0319) [inline src/types/PlayerBlockAction.php]:
```suggestion
		if(match($this->actionType){
```

File: src/types/PlayerBlockAction.php [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+namespace pocketmine\network\mcpe\protocol\types;
+
+use pocketmine\network\mcpe\protocol\serializer\PacketSerializer;
+
+/** This is used for PlayerAuthInput packet when the flags include PERFORM_BLOCK_ACTIONS */
+final class PlayerBlockAction{
+
+	public const START_BREAK = 0;
+	public const ABORT_BREAK = 1;
```
Comment (Prim69) [inline src/types/PlayerBlockAction.php:30]: Reminder for me to remove these constants and use the ones in PlayerAction when I wake up

**Comment by Prim69** [comment]
Everything so far has been resolved now

**Review by dktapps — CHANGES_REQUESTED** [review]

File: src/PlayerAuthInputPacket.php [hunk 7]
```diff
@@ -47,14 +52,22 @@ class PlayerAuthInputPacket extends DataPacket implements ServerboundPacket{
...
 	private int $tick;
 	private Vector3 $delta;
+	private ?UseItemTransactionData $itemInteractionData = null;
+	private ?ItemStackRequest $itemStackRequest = null;
+	/** @var PlayerBlockAction[]|null */
+	private ?array $blockActions = null;
+
+	private ?int $requestId = null;
+	/** @var InventoryTransactionChangedSlotsHack[]|null */
+	private ?array $requestChangedSlots = null;
```
[the remark this answers is not in the document]
Comment (Prim69) [reply src/PlayerAuthInputPacket.php:62]: Should I put the UseItemTransactionData in the constructor of the parent class or just set it to new UseItemTransactionData() as it's currently done (seen on L203)?

[the remark this answers is not in the document]
Comment (dktapps) [reply src/PlayerAuthInputPacket.php:62]: I have no idea what you mean.

[the remark this answers is not in the document]
Comment (Prim69) [reply src/PlayerAuthInputPacket.php:62]:
```php
final class ItemInteractionData{
	public function __construct(UseItemTransactionData $transactionData, int $requestId, array $requestChangedSlots){
		$this->requestId = $requestId;
		$this->transactionData = $transactionData;
		$this->requestChangedSlots = $requestChangedSlots;
	}
}
```
vs this
```php
final class ItemInteractionData{
	public function __construct(int $requestId, array $requestChangedSlots){
		$this->requestId = $requestId;
		$this->transactionData = new UseItemTransactionData();
		$this->requestChangedSlots = $requestChangedSlots;
	}
}
```
I'm not sure if there would be any circumstances that would require the first option tho

[the remark this answers is not in the document]
Comment (dktapps) [reply src/PlayerAuthInputPacket.php:62]: You should do the first way. The second way allows for accidents and it doesn't really make any sense.

File: src/PlayerAuthInputPacket.php [hunk 8]
```diff
@@ -159,9 +232,28 @@ protected function encodePayload(PacketSerializer $out) : void{
...
+			$out->putVarInt($this->requestId);
+			if($this->requestId !== 0){
+				$out->putUnsignedVarInt(count($this->requestChangedSlots));
+				foreach($this->requestChangedSlots as $changedSlot){
+					$changedSlot->write($out);
+				}
+			}
+			$this->itemInteractionData->encode($out);
+		}
+		if($this->hasFlag(PlayerAuthInputFlags::PERFORM_ITEM_STACK_REQUEST)){
```
[the remark this answers is not in the document]
Comment (dktapps) [reply src/PlayerAuthInputPacket.php:245]: You need to use null checks here instead of checking the flags.

File: src/PlayerAuthInputPacket.php [hunk 9]
```diff
@@ -202,10 +205,10 @@ protected function decodePayload(PacketSerializer $in) : void{
 			$this->itemInteractionData = new ItemInteractionData($requestId, $requestChangedSlots, new UseItemTransactionData());
 			$this->itemInteractionData->getTransactionData()->decode($in);
 		}
-		if($this->hasFlag(PlayerAuthInputFlags::PERFORM_ITEM_STACK_REQUEST)){
+		if($this->itemStackRequest !== null){
 			$this->itemStackRequest = ItemStackRequest::read($in);
 		}
-		if($this->hasFlag(PlayerAuthInputFlags::PERFORM_BLOCK_ACTIONS)){
+		if($this->blockActions !== null){
```
Comment (dktapps) [inline src/PlayerAuthInputPacket.php:211]: Not these.

File: src/types/PlayerBlockAction.php [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+final class PlayerBlockAction{
+
+	public int $actionType;
+	public BlockPosition $blockPos;
+	public int $face;
+
+	public function read(PacketSerializer $in) : void{
+		$this->actionType = $in->getVarInt();
+		if(match($this->actionType){
+			PlayerAction::ABORT_BREAK, PlayerAction::START_BREAK, PlayerAction::CRACK_BREAK, PlayerAction::PREDICT_DESTROY_BLOCK, PlayerAction::CONTINUE_DESTROY_BLOCK => true,
```
Comment (dktapps) [inline src/types/PlayerBlockAction.php:36]: This looks awful.

Comment (Prim69) [reply src/types/PlayerBlockAction.php:36]: The other 2 options are
```php
if($this->actionType !== PlayerAction::STOP_BREAK && $this->actionType >= PlayerAction::START_BREAK){}
```
or one with a bunch of `or`s

Comment (dktapps) [reply src/types/PlayerBlockAction.php:36]: I think a `switch` would be good enough here, to be honest.

**Review by dktapps — APPROVED** [review]

**Pull request merged**